### PR TITLE
Add `mas` module for installing applications via the Mac App Store

### PR DESCRIPTION
**Pull request by mheap**
Currently, only the `install` command is available, and you must provide IDs
when installing applications.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
mas

##### ANSIBLE VERSION
```
ansible 2.3.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```

##### SUMMARY
This adds a new `mas` module that can be used to install packages on OSX from the Mac App Store. The module requires that `mas` is preinstalled.

You can currently only install packages by ID, but there is scope to add a `mas_facts` module to allow searching via the MAS API to populate a lookup from application name to ID. You can see a discussion I had with @dfederlein around this [on Reddit](https://www.reddit.com/r/ansible/comments/4y6smd/new_module_mas_install_apps_from_the_mac_app_store/)

### CODE REVIEW

**Comment by mheap** [comment]
I could use some guidance here. I had `except Exception as e` in the code which failed `python2.4` compilation. After updating it to the legacy style exception handler that check passes, but `python3` fails.

What's the recommended way to proceed?

**Comment by mattclay** [comment]
@mheap Here's an example of how to handle exceptions in python 2.4+:

https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/files/file.py#L286

**Comment by mheap** [comment]
@mattclay Thanks for pointing me in the correct direction. I've solved all of the issues raised in CI. Let me know if there's anything else I need to do

**Comment by mheap** [comment]
Any update on this?

@DJMuggs @JoergFiedler @KimNorgaard @L2G @barnabycourt @berenddeboer @brian-brazil @brontitall @dermute @eest @evgkrsk @FlossWare @hacosta @ignatenkobrain @indrajitr @jcftang @jirutka @jtyr @jvantuyl @kbrebanov @matze @mavit @melodous @pmakowski @robinro @skinp @szinck @tdtrask @troy2914 @vincentvdk @vritant @wltjr

**Comment by mattclay** [comment]
bot_status

**Comment by Akasurde** [comment]
@mheap Are you still working on this ? needs_info

**Comment by mheap** [comment]
I’d left it as it didn’t seem to be getting any attention. I’m happy to clean it up if it’ll help us progress

File: lib/ansible/modules/packaging/os/mas.py
Comment (Akasurde) [inline lib/ansible/modules/packaging/os/mas.py]: Replace line 7 to 20 with 

```
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)

from __future__ import absolute_import, division, print_function
__metaclass__ = type

```

Comment (Akasurde) [inline lib/ansible/modules/packaging/os/mas.py]:
```
# Copyright: (c) 2017, Michael Heap
```

File: lib/ansible/modules/packaging/os/mas.py
Comment (Akasurde) [inline lib/ansible/modules/packaging/os/mas.py]:
```
ANSIBLE_METADATA = {
    'status': ['preview'],
    'supported_by': 'community',
    'metadata_version': '1.1'
}
```

File: lib/ansible/modules/packaging/os/mas.py
Comment (Akasurde) [inline lib/ansible/modules/packaging/os/mas.py]: s/2.4/2.6/

File: lib/ansible/modules/packaging/os/mas.py
Comment (Akasurde) [inline lib/ansible/modules/packaging/os/mas.py]: Period at end.

File: lib/ansible/modules/packaging/os/mas.py
Comment (Akasurde) [inline lib/ansible/modules/packaging/os/mas.py]: Period at end.

File: lib/ansible/modules/packaging/os/mas.py
Comment (Akasurde) [inline lib/ansible/modules/packaging/os/mas.py]: How about 'absent' as state, which make this module complete.

File: lib/ansible/modules/packaging/os/mas.py
Comment (Akasurde) [inline lib/ansible/modules/packaging/os/mas.py]: Use `get_bin_path` instead of this function.

File: lib/ansible/modules/packaging/os/mas.py
Comment (Akasurde) [inline lib/ansible/modules/packaging/os/mas.py]: Add absent as well.

File: lib/ansible/modules/packaging/os/mas.py
Comment (Akasurde) [inline lib/ansible/modules/packaging/os/mas.py]: get_exception() is deprecated as Ansible does not support Python 2.4 version.

File: lib/ansible/modules/packaging/os/mas.py
Comment (Akasurde) [inline lib/ansible/modules/packaging/os/mas.py]: Use `to_native`

**Review by Akasurde — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/packaging/os/mas.py
Comment (mheap) [reply lib/ansible/modules/packaging/os/mas.py]: The underlying `mas` tool does not support removals. only installations

Comment (Akasurde) [reply lib/ansible/modules/packaging/os/mas.py]: Could you please add the note in documentation mentioning this limitation ?

**Comment by mheap** [comment]
@Akasurde I've addressed all of your comments. Thanks for `get_bin_path`, I didn't know it existed!

Is there anything else I need to do before this can be merged?

**Review by Akasurde — APPROVED** [review]

**Comment by mheap** [comment]
@Akasurde Just updated to 2.8 and all CI is passing. Can you see any other changes I need to make?

**Comment by lukasbestle** [comment]
This module would also be very useful to me. Are there any news on its current state, i.e. is there anything missing to merge this?

@mheap The `mas` utility now supports `upgrade` and `uninstall` as well, so the module could accept `latest` and `absent` as states. In case this module will get merged, I could help with the implementation of those features if you don't have time. Would be my first contribution to an Ansible module though.

**Comment by mheap** [comment]
@lukasbestle If you'd like to try implementing the features, I'd be happy to review it for you

**Comment by lukasbestle** [comment]
@mattclay @Akasurde Would this PR be merged with the current state and/or with the additional states implemented?

**Comment by mattclay** [comment]
Since the underlying tool now supports it, the initial version should include `latest` and `absent` since those are features normally expected for Ansible package management modules.

**Comment by lukasbestle** [comment]
Thanks, I'm working on it and will send a PR to @mheap's PR branch.

**Comment by lukasbestle** [comment]
@mattclay @Akasurde This PR is now ready for review again. The new `mas` module now has all the features of the underlying `mas-cli` tool that are relevant for automation. I have also improved the documentation and code structure.

File: lib/ansible/modules/packaging/os/mas.py
Comment (felixfontein) [inline lib/ansible/modules/packaging/os/mas.py]:
```suggestion
    - Installs, uninstalls and updates macOS applications from the Mac App Store using the C(mas-cli).
```
(See https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#linking-within-module-documentation)

File: lib/ansible/modules/packaging/os/mas.py
Comment (felixfontein) [inline lib/ansible/modules/packaging/os/mas.py]:
```suggestion
    - "mas-cli (U(https://github.com/mas-cli/mas)) 1.5.0+ available as C(mas) in the bin path"
```

File: lib/ansible/modules/packaging/os/mas.py
Comment (felixfontein) [inline lib/ansible/modules/packaging/os/mas.py]: I wouldn't exactly call a dictionary `list`.

File: lib/ansible/modules/packaging/os/mas.py
Comment (felixfontein) [inline lib/ansible/modules/packaging/os/mas.py]:
```suggestion
    for app in sorted(set(apps)):
```
This way, you won't have trouble if apps show up more than once in the list (in which case `is_installed` and `is_outdated` can return wrong information), and the order is still deterministic also for older Python versions.

File: lib/ansible/modules/packaging/os/mas.py
Comment (felixfontein) [inline lib/ansible/modules/packaging/os/mas.py:229]: Here, `outdated` is a `dict` (or `list` in mheap/ansible#3).

File: lib/ansible/modules/packaging/os/mas.py
Comment (felixfontein) [inline lib/ansible/modules/packaging/os/mas.py]: Here, `self.count_upgrade` is an int, and `outdated` (as above) a `dict` resp. `list`.

Comment (lukasbestle) [reply lib/ansible/modules/packaging/os/mas.py]: You are right! Fixed it in https://github.com/mheap/ansible/pull/3.

**Review by felixfontein — APPROVED** [review]
LGTM
(without any `mas` knowledge though)

**Comment by lukasbestle** [comment]
@mattclay @Akasurde Ping :)

**Comment by lukasbestle** [comment]
Does this have a chance to get merged anytime soon or even just reviewed by the core team? @Akasurde @mattclay?

To be honest I wouldn't have contributed if I had known it would be such a lengthy process. And also my motivation to contribute more improvements and features is near zero at the moment. That's very unfortunate as I really like Ansible and would love to help moving it forward... 😿

**Comment by felixfontein** [comment]
@lukasbestle I think there is. I'll try to ping @mattclay and @Akasurde again the next days to see whether they have objections to this being merged. Currently, the core team is pretty busy preparing the migration to collections, so they probably won't be able to do more than that, but assuming they are ok with it I'll simply merge this.

> To be honest I wouldn't have contributed if I had known it would be such a lengthy process. And also my motivation to contribute more improvements and features is near zero at the moment. That's very unfortunate as I really like Ansible and would love to help moving it forward... crying_cat_face

This unfortunately happened a lot over the last years. The main problem is that the core team is way too small to take care of both the core *and* the thousands (!) of community modules. And if nobody from the community steps up to help getting such a PR merged, not much will happen. (Unfortunately.)

**Comment by mattclay** [comment]
I don't have objections to this being merged, although it would be nice to see an integration test for it -- even though we won't be able to run it in CI due to the credentials requirement. I haven't reviewed the module, only provided feedback on a few questions.

**Comment by felixfontein** [comment]
For creating a simple integration test, do something like this:
1. Create test/integration/targets/mas/aliases with content
```
needs/root
skip/freebsd
skip/rhel
skip/docker
unsupported
```
2. Create test/integration/targets/mas/tasks/main.yml which contains a few tasks which use this module. Do something like install something, install it again (with an assert that the second task didn't change), remove it (assert that change occured), remove it again (assert that no change occured), and maybe a call with `upgrade_all: yes`.

Please also try to run that test (`ansible-test integration mas`) and check that it works.

**Comment by mattclay** [comment]
Note that the `skip/*` aliases aren't needed since `unsupported` skips the tests in CI completely.

**Comment by gundalow** [comment]
I think this is good as is, so merging

**Comment by felixfontein** [comment]
@lukasbestle @mheap it would still be nice to have tests, but let's do that after modules have been moved to collections (i.e. in a couple of weeks). If you want to do it right now, that's also fine, then there's a good chance to get it merged before the move :)

(If you create a PR, please CC me.)

**Comment by lukasbestle** [comment]
@felixfontein I'm currently working on the integration test. How can I make `ansible-test` ask me for the become password?

I'm trying with:

```
ansible-test integration --allow-unsupported --allow-root mas
```

...but it fails with:

```
fatal: [testhost]: FAILED! => {"changed": false, "module_stderr": "sudo: a password is required\n", "module_stdout": "", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 1}
```

**Comment by felixfontein** [comment]
Hmm, good question. For some other tests that need root, I ran them locally not with `ansible-test`, but by putting a playbook `run-tests-locally.yml` into `test/integration/targets/xxx/` with the following content:
```
---
- hosts: localhost
  tasks:
  - include_tasks: tasks/main.yml
    vars:
      role_path: '.'
```
(with potential setup tasks before the `include_tasks`) and executed that one. Then I can also pass all regular become-related arguments :)

**Comment by lukasbestle** [comment]
Thanks! PR incoming.

**Pull request merged**
